### PR TITLE
feat: Support only Gen2 deployment type for cloudspace resource

### PR DESCRIPTION
**Pull request by TheNilesh**
Modify the attribute validation rule in `provider_code_spec.json` to disallow the `gen1` value.

### CODE REVIEW

File: internal/provider/cloudspace_resource.go [hunk 1]
```diff
@@ -63,6 +63,23 @@ func (r *cloudspaceResource) Configure(ctx context.Context, req resource.Configu
 }
 
 func (r *cloudspaceResource) ModifyPlan(ctx context.Context, req resource.ModifyPlanRequest, resp *resource.ModifyPlanResponse) {
+	if req.Plan.Raw.IsNull() {
+		// resource is being destroyed
```
Comment (indradhanush) [inline internal/provider/cloudspace_resource.go]: Should we log this?

File: internal/provider/cloudspace_resource.go [hunk 2]
```diff
@@ -63,6 +63,23 @@ func (r *cloudspaceResource) Configure(ctx context.Context, req resource.Configu
...
+		return
+	}
+	if req.State.Raw.IsNull() && !req.Plan.Raw.IsNull() {
```
Comment (indradhanush) [inline internal/provider/cloudspace_resource.go]: > !req.Plan.Raw.IsNull()

This is not needed here right? We've already returned early above if `Plan.Raw` is null? So here its always not null?

File: provider_code_spec.json [hunk 3]
```diff
@@ -437,12 +438,9 @@
 										"imports": [
 											{
 												"path": "github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
-											},
-											{
-												"path": "regexp"
 											}
 										],
-										"schema_definition": "stringvalidator.RegexMatches(regexp.MustCompile(`^gen[12]$`), \"deployment_type must be gen1 (or) gen2\")"
+										"schema_definition": "stringvalidator.OneOf(\"gen1\", \"gen2\")"
```
Comment (indradhanush) [inline provider_code_spec.json]: This should only contain `gen2`?

File: internal/provider/resource_cloudspace/cloudspace_resource_gen.go [hunk 4]
```diff
@@ -75,16 +75,17 @@ func CloudspaceResourceSchema(ctx context.Context) schema.Schema {
...
+				Description:         "Specifies the deployment type for the cloudspace (Only gen2 is allowed value).",
+				MarkdownDescription: "Specifies the deployment type for the cloudspace (Only gen2 is allowed value).",
+				DeprecationMessage:  "Support will be limited to Gen2 cloudspaces, rendering this field obsolete.",
 				PlanModifiers: []planmodifier.String{
 					stringplanmodifier.RequiresReplace(),
 					stringplanmodifier.UseStateForUnknown(),
 				},
 				Validators: []validator.String{
-					stringvalidator.RegexMatches(regexp.MustCompile(`^gen[12]$`), "deployment_type must be gen1 (or) gen2"),
+					stringvalidator.OneOf("gen1", "gen2"),
```
Comment (indradhanush) [inline internal/provider/resource_cloudspace/cloudspace_resource_gen.go]: We only want to support `gen2` right? Why are we still supporting `gen1` as a valid input here?

**Review by indradhanush — COMMENTED** [review]
General comments as I haven't yet interacted with this repo in depth yet.

Will recommend requesting a review from another contributor.

File: provider_code_spec.json (hunk 3 above)
Comment (TheNilesh) [reply provider_code_spec.json]: Existing Gen1 cloudspaces will cease functioning (updates will fail) if validations are enforced. Therefore, both are permitted. Validations are applied in `ModifyPlan()`, where a distinction between Create and Update operations can be made.

Comment (TheNilesh) [reply provider_code_spec.json]: I see the reasoning behind restricting Gen1 through attribute validations, as users can still manage Gen1 cloudspaces with an older provider version.

File: internal/provider/cloudspace_resource.go (hunk 1 above)
Comment (TheNilesh) [reply internal/provider/cloudspace_resource.go]: This change has been reverted, so logging is no longer necessary.

File: internal/provider/resource_cloudspace/cloudspace_resource_gen.go (hunk 4 above)
Comment (TheNilesh) [reply internal/provider/resource_cloudspace/cloudspace_resource_gen.go]: Corrected

File: internal/provider/cloudspace_resource.go (hunk 2 above)
Comment (TheNilesh) [reply internal/provider/cloudspace_resource.go]: This change has been reverted.

**Review by ansalamdaniel — APPROVED** [review]
LGTM

**Review by pshanbhag — COMMENTED** [review]
Do you need to update [terraform-provider-spot](https://github.com/rackerlabs/terraform-provider-spot/tree/0d90d480ba4a770db807b7770afb0b8414dec0de)/[examples](https://github.com/rackerlabs/terraform-provider-spot/tree/0d90d480ba4a770db807b7770afb0b8414dec0de/examples)/[cloudspace](https://github.com/rackerlabs/terraform-provider-spot/tree/0d90d480ba4a770db807b7770afb0b8414dec0de/examples/cloudspace/main.tf) and [terraform-provider-spot](https://github.com/rackerlabs/terraform-provider-spot/tree/0d90d480ba4a770db807b7770afb0b8414dec0de)/[docs](https://github.com/rackerlabs/terraform-provider-spot/tree/0d90d480ba4a770db807b7770afb0b8414dec0de/docs)/[data-sources](https://github.com/rackerlabs/terraform-provider-spot/tree/0d90d480ba4a770db807b7770afb0b8414dec0de/docs/data-sources/cloudspace.md)?

**Pull request merged**